### PR TITLE
feat(ml): IncrementalPredictor

### PR DESCRIPTION
**Pull request by JovanVeljanoski**
This PR adds a `veax.ml.sklearn.IncrementalPredictor` class that wraps any `scikit-learn` estimator (a.k.a predictor) that has the `.partial_fit` method. 

This means that we can send batches of data from `vaex` to such a predictor, while controlling the memory needed. 

Each batch can be "seen" by the predictor multiple times (epochs), and optionally it can be shuffled.

- [x] Implementation
- [x] Unit-tests
- [ ] Documentation
- [ ] Code review

### CODE REVIEW

File: packages/vaex-ml/vaex/ml/sklearn.py [hunk 1]
```diff
@@ -95,3 +95,141 @@ def fit(self, df, target, **kwargs):
         X = df[self.features].values
         y = df.evaluate(target)
         self.model.fit(X=X, y=y, **kwargs)
+
+
+@vaex.serialize.register
+@generate.register
+class IncrementalPredictor(state.HasState):
+    '''This class wraps any scikit-learn estimator (a.k.a predictions) that has
+    a `.partial_fit` method, and makes it a vaex pipeline object.pipeline
```
Comment (maartenbreddels) [inline packages/vaex-ml/vaex/ml/sklearn.py]:
```suggestion
    a `.partial_fit` method, and makes it a vaex pipeline object.
```

File: packages/vaex-ml/vaex/ml/sklearn.py [hunk 2]
```diff
@@ -95,3 +95,141 @@ def fit(self, df, target, **kwargs):
...
+
+    By wrapping "on-line" scikit-learn estimators with this class, they becoes a vaex
```
Comment (maartenbreddels) [inline packages/vaex-ml/vaex/ml/sklearn.py]:
```suggestion
    By wrapping "on-line" scikit-learn estimators with this class, they become a vaex
```

File: packages/vaex-ml/vaex/ml/sklearn.py [hunk 3]
```diff
@@ -95,3 +95,141 @@ def fit(self, df, target, **kwargs):
...
+    pipeline object. Thus, they can take full advantage of the serialization and
+    pipeline system of vaex. While the underlying estimator need to call the
+    `.partial_fit` method, this class contains the standard `.fit` method, and
+    the rest happens behind the scenes. One can also iterate over the data
+    multiple times (epochs), and optionally shuffle each batch before it is sent
+    to the estimator. The `predict` method returns a numpy, while the `transform`
```
Comment (maartenbreddels) [inline packages/vaex-ml/vaex/ml/sklearn.py]:
```suggestion
    to the estimator. The `predict` method returns a numpy array, while the `transform`
```

File: packages/vaex-ml/vaex/ml/sklearn.py [hunk 4]
```diff
@@ -95,3 +95,141 @@ def fit(self, df, target, **kwargs):
...
+    method adds the prediction as a virtual column to a vaex DataFrame.
```
Comment (maartenbreddels) [inline packages/vaex-ml/vaex/ml/sklearn.py]: this is general docs, should not be put here right?

File: packages/vaex-ml/vaex/ml/sklearn.py [hunk 5]
```diff
@@ -95,3 +95,141 @@ def fit(self, df, target, **kwargs):
...
+
+    Note that the `.fit` method will use as much memory as needed to copy one
+    batch of data, while the `.predict` method makes a full memory copy.
+    The `transform` method is evaluated lazily, and no memory copies are made.
+
+    The scikit-learn estimators themselves are not modified at all, they are
+    taken from your local installation of scikit-learn.
+
```
Comment (maartenbreddels) [inline packages/vaex-ml/vaex/ml/sklearn.py]: I think the wording is a bit odd, maybe we can completely remove it, or rephrase it simpler. "Note, we are using normal sklearn without modifications here."

File: packages/vaex-ml/vaex/ml/sklearn.py [hunk 6]
```diff
@@ -95,3 +95,141 @@ def fit(self, df, target, **kwargs):
...
+        # For reference: tqdm version was - would be nice to have an informative progress bar
+        # for i in tqdm(range(num_batches), leave=True, desc='Training in batches...'):
+
+        # Main loop: pass on batches to the model
+        for batch in range(num_batches):
+            # Get data for this batch
+            X = df[index_min:index_max][self.features].values
+            y = df[index_min:index_max][target].values
+            shuffle_index = np.arange(len(X))
```
Comment (maartenbreddels) [inline packages/vaex-ml/vaex/ml/sklearn.py]:
```suggestion
```

File: packages/vaex-ml/vaex/ml/sklearn.py [hunk 7]
```diff
@@ -95,3 +95,141 @@ def fit(self, df, target, **kwargs):
...
+        '''Fit the IncrementalPredictor to the DataFrame.
+
+        :param df: A vaex DataFrame containing the features on which to train the model.
+        :param target: The name of the column containing the target variable.
+        '''
+        # Check whether the model is appropriate
+        assert hasattr(self.model, 'partial_fit'), 'The model must have a `.partial_fit` method.'
+
+        # Number of batches to send (or training rounds)
+        num_batches = np.int(np.ceil(len(df) / self.batch_size))
```
Comment (maartenbreddels) [inline packages/vaex-ml/vaex/ml/sklearn.py]:
```suggestion
        N_total = len(df)
        num_batches = (N_total + self.batch_size -1) // self.batch_size  # round up int
```

File: packages/vaex-ml/vaex/ml/sklearn.py [hunk 8]
```diff
@@ -95,3 +95,141 @@ def fit(self, df, target, **kwargs):
...
+                    np.random.shuffle(shuffle_index)
+                    X = X[shuffle_index]
+                    y = y[shuffle_index]
+
+                # train the model
+                self.model.partial_fit(X, y)
+
+            # update the slicing indices to be ready for the next batch
+            index_min += self.batch_size
+            index_max += self.batch_size
```
Comment (maartenbreddels) [inline packages/vaex-ml/vaex/ml/sklearn.py]:
```suggestion
            index_max += self.batch_size
            index_max = min(N_total, index_max)  # clip upper value
```
Is this not triggered by the test?

File: packages/vaex-ml/vaex/ml/sklearn.py [hunk 9]
```diff
@@ -95,3 +95,141 @@ def fit(self, df, target, **kwargs):
...
+
+    def predict(self, df):
+        '''Get an in-memory numpy array with the predictions of the SKLearnPredictor.self
+
+        :param df: A vaex DataFrame, containing the input features.
+        :returns: A in-memory numpy array containing the SKLearnPredictor predictions.
+        :rtype: numpy.array
+        '''
+        data = df[self.features].values
+        return self.model.predict(data)
```
Comment (maartenbreddels) [inline packages/vaex-ml/vaex/ml/sklearn.py]: why don't do here
```
return self.transform(df)[self.prediction_name].values
```
should be much more efficient right? should also update the docs for that

File: tests/ml/sklearn_test.py [hunk 10]
```diff
@@ -160,3 +160,56 @@ def test_sklearn_estimator_classification_validation():
...
+def test_sklearn_incremental_predictor_serialize(tmpdir):
+    df = vaex.example()
+    df_train, df_test = df.ml.train_test_split(test_size=0.1, verbose=False)
+
+    features = df_train.column_names[:6]
+    target = 'FeH'
+
+    incremental = IncrementalPredictor(model=SGDRegressor(),
+                                       features=features,
+                                       batch_size=10_000,
```
Comment (maartenbreddels) [inline tests/ml/sklearn_test.py]: Make batchsize an odd number and see if you can check from the skmodel how many samples it has seen.

File: packages/vaex-ml/vaex/ml/sklearn.py (hunk 5 above)
Comment (JovanVeljanoski) [reply packages/vaex-ml/vaex/ml/sklearn.py]: I just copy-pasted it from the regular SKLearnPredictor, but we can change the wording

File: packages/vaex-ml/vaex/ml/sklearn.py (hunk 9 above)
Comment (JovanVeljanoski) [reply packages/vaex-ml/vaex/ml/sklearn.py]: we could indeed!

File: tests/ml/sklearn_test.py (hunk 10 above)
Comment (JovanVeljanoski) [reply tests/ml/sklearn_test.py]: Unfortunately the sklearn models do not keep track on the number of samples they have seen.. however I can test this manually..

File: packages/vaex-ml/vaex/ml/sklearn.py (hunk 4 above)
Comment (JovanVeljanoski) [reply packages/vaex-ml/vaex/ml/sklearn.py]: Yeah, but I don't find it overly long, and it might be helpful to new users.

Comment (maartenbreddels) [reply packages/vaex-ml/vaex/ml/sklearn.py]: But my point is that general info should not be put everywhere, but in one place to avoid maintenance burden, just an advice.

File: tests/ml/sklearn_test.py (hunk 10 above)
Comment (maartenbreddels) [reply tests/ml/sklearn_test.py]: You could use mock to wrap the method, e.g.:
https://stackoverflow.com/questions/18869141/using-python-mock-to-spy-on-calls-to-an-existing-object

Comment (JovanVeljanoski) [reply tests/ml/sklearn_test.py]: Added an additional test uses `mock` to count the number of calls to the underlying `partial_fit` method to the model.

File: tests/ml/sklearn_test.py [hunk 11]
```diff
@@ -213,3 +214,25 @@ def test_sklearn_incremental_predictor_serialize(tmpdir):
...
+    num_batches = (N_total + batch_size - 1) // batch_size
+
+    incremental = IncrementalPredictor(model=mock.Mock(),
+                                       features=features,
+                                       batch_size=batch_size,
+                                       num_epochs=num_epochs,
+                                       shuffle=False,
+                                       prediction_name='pred')
+    incremental.fit(df=df_train, target=target)
+    assert incremental.model.partial_fit.call_count == num_batches * num_epochs
```
Comment (maartenbreddels) [inline tests/ml/sklearn_test.py]: This is a pretty trivial assertion. I am more worried about the # of rows being passed being equal to len(df_train).
